### ISSUE TITLE
Filehandle limits in init scripts lead to extremely poor python performance

### ISSUE THREAD

**Issue by shanemhansen** [docker/docker#9876]
Let me preface this by saying this isn't strictly a bug report, but docker and python interact in a way that can give extremely poor performance.

tl;dr - certain types of python subprocess execution inside docker takes seconds and causes extremely high cpu usage.

I first noticed this bug while running salt in docker because salt shells out to `ip addr` and I found that my machine was maxing out 4 cores to do so. https://github.com/saltstack/salt/issues/18569

I'll spare you the whole debugging session (it's in the linked bug if you're interested), but basically this existing python bug: http://bugs.python.org/issue1663329 is triggered by this init script: https://github.com/docker/docker/blob/master/contrib/init/upstart/docker.conf#L5

# Steps to reproduce

When this command is ran:

```
time docker run -i -t ubuntu:12.04  python -c "import subprocess;subprocess.Popen(['/bin/ls'], close_fds=True).wait()"
```

# Root cause

python to closes over 500k invalid file descriptors, which sucks. What sucks more is that python constructs an exception and throws it and catches it half a million times.

# Potential Impact

Existing python programs are unlikely to change the flag that causes this problem (close_fds=True) because not giving a child process access to your filehandles is a security feature. Just by doing a casual grep on ubuntu (I assume one of the most common docker images ran) I can see that builtin packages like lsb_release.py and debconf.py use the close_fds=True flag.

I don't know the best fix for this, I don't want people to get the impression (deserved or not) that running code in docker is slow, but obviously it's not really docker's fault that python takes a really long time to shell out if the filehandle limit is high. It's possible this could be addressed by documentation or by changing the default open filehandle limits in the init scripts.

**Comment by shanemhansen** [comment]
So far I've seen this in python 2.6 and 2.7 (2.6.6 and 2.7.3).

**Comment by jfrazelle** [comment]
what do you offer as a solution?

**Comment by shanemhansen** [comment]
Decrease the default filehandle limit specified in the [contrib init files](https://github.com/docker/docker/blob/master/contrib/init/upstart/docker.conf) down from half a million and document how to bump it back up.

**Comment by tianon** [comment]
Wouldn't having first class support in Docker itself for ulimit management
be a better solution to this issue?

**Comment by shanemhansen** [comment]
I think having ulimit management in docker would be a good idea. The ulimit values appropriate for running the docker daemon aren't necessarily appropriate for pid 1 in the container.

**Comment by thaJeztah** [comment]
@tianon are you referring to https://github.com/docker/docker/pull/9437?

**Comment by jfrazelle** [comment]
closing as dup of https://github.com/docker/docker/issues/4717
